vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java

```diff
@@ -195,9 +318,73 @@ private void refresh() {
       piece = null;
     }
     slot.setPiece(piece);
+
+    if (piece != null) {
+      // Check the size of the image and re-size image and/or panel if required
+      resizeSlotPanel();
+    }
+
+    slotPanel.revalidate();
+    repack();
+    slot.getComponent().repaint();
+  }
+
+  /**
+   *
+   * Re-calculate how to layout the left hand Piece display panel.
+   *
+   * 1. Always try and show the entire image 100% scale
+   * 2. If the image can't fit in the available space, scale it until it can
+   * 3. If the image size is < max split size, then set the divider size = image size to now waste space
+   *
+   */
+  private void resizeSlotPanel() {
+
+    // Reset the piece to 100% scale to get the actual size.
+    final double oldScale = slot.getScale();
+    slot.setScale(1.0);
+    final Dimension newSlotSize = new Dimension(slot.getPiece().boundingBox().width, slot.getPiece().boundingBox().height);
```

Comment:
Getting the bb could be expensive, maybe get it and store it rather than getting it twice?

Reply:
Fixed